Kernel/Devices/HID/I8042Controller.cpp

```diff
@@ -31,8 +31,30 @@ UNMAP_AFTER_INIT I8042Controller::I8042Controller()
 {
 }
 
+UNMAP_AFTER_INIT bool I8042Controller::check_existence()
+{
+    {
+        SpinlockLocker lock(m_lock);
+        // Note: Perform controller self-test before touching the controller
+        // Try to probe the controller for 5 times and give up if nothing
+        // responded.
+        for (int attempt = 0; attempt < 5; attempt++) {
+            do_write(I8042Port::Command, I8042Command::TestPS2Controller);
+            if (do_read(I8042Port::Buffer) == I8042Response::ControllerTestPassed)
+                return true;
+            // Note: Wait 50 microseconds in case the controller couldn't respond
+            IO::delay(50);
+        }
+        dbgln("I8042: Controller self test failed");
```

Comment:
last nitpick: It's true that the self test failed, but I would debug something more meaningful here, like "could not find any controllers" etc